File: packages/ink/src/text-normalisation/normalise-text.algorithm.ts
import { foldConfusables } from './confusables.mapper'
import { foldDiacritics } from './diacritics.mapper'

/**
 * Text made comparable: the differences OCR introduces that say nothing about
 * the document are removed, and nothing else.
 *
 * Each step is a judgement about which noise to forgive, so each can be turned
 * off. In order:
 *
 * 1. Unicode NFKC - ligatures (`ﬁ`), full-width and compatibility forms.
 * 2. Typography - curly quotes, dashes, ellipses and odd spaces to plain ASCII.
 * 3. De-hyphenation - `informa-` + newline + `tion` is `information`, or every
 *    wrapped word counts as an edit.
 * 4. Diacritics - `é` to `e`, `Æ` to `AE`.
 * 5. Case.
 * 6. Noise - tokens with no letter, digit or currency sign: table rules read
 *    as `|`, underlines as `___`, specks as `.` or `~`.
 * 7. Punctuation (off by default) - it carries real signal: `1,250.00` is not
 *    `125000`.
 * 8. OCR confusables (off by default) - `0/o`, `1/l/i`, `5/s`, `rn/m`. Folding
 *    them forgives exactly the substitutions a forger would make.
 * 9. Whitespace collapsed and trimmed.
 */

export interface NormaliseOptions {
  nfkc?:             boolean
  typography?:       boolean
  dehyphenate?:      boolean
  diacritics?:       boolean
  caseFold?:         boolean
  dropNoise?:        boolean
  stripPunctuation?: boolean
  confusables?:      boolean
}

export const DEFAULT_NORMALISE: Required<NormaliseOptions> = {
  nfkc:             true,
  typography:       true,
  dehyphenate:      true,
  diacritics:       true,
  caseFold:         true,
  dropNoise:        true,
  stripPunctuation: false,
  confusables:      false,
}

const LINE_END_HYPHEN = /(\p{L})[-\u{AD}\u{2010}\u{2011}][\t ]*\r?\n[\t ]*(?=\p{Ll})/gu
const TYPOGRAPHY: ReadonlyArray<readonly [RegExp, string]> = [
  // single quotes and prime, double quotes and double prime
  [/[\u{2018}\u{2019}\u{201A}\u{201B}\u{2032}]/gu, '\''],
  [/[\u{201C}\u{201D}\u{201E}\u{201F}\u{2033}]/gu, '"'],
  // hyphens, figure dash, en and em dashes, horizontal bar, minus
  [/[\u{2010}\u{2011}\u{2012}\u{2013}\u{2014}\u{2015}\u{2212}]/gu, '-'],
  [/\u{2026}/gu, '...'],
  // no-break and typographic spaces
  [/[\u{A0}\u{2000}-\u{200A}\u{202F}\u{205F}\u{3000}]/gu, ' '],
  // zero-width space and non-joiner, word joiner, byte-order mark
  [/[\u{200B}\u{200C}\u{2060}\u{FEFF}]/gu, ''],
  // zero-width joiner, on its own: in a character class it reads as joining its neighbours
  [/\u{200D}/gu, ''],
]
const MEANINGFUL = /[\p{L}\p{N}\p{Sc}]/u
const PUNCTUATION = /\p{P}+/gu
const WHITESPACE = /\s+/gu

export function normaliseText (text: string, options: NormaliseOptions = {}): string {
  const o = { ...DEFAULT_NORMALISE, ...options }
  let out = text

  if (o.nfkc) out = out.normalize('NFKC')
  if (o.typography) for (const [pattern, replacement] of TYPOGRAPHY) out = out.replaceAll(pattern, () => replacement)
  if (o.dehyphenate) out = out.replaceAll(LINE_END_HYPHEN, '$1')
  if (o.diacritics) out = foldDiacritics(out)
  if (o.caseFold) out = out.toLowerCase()
  if (o.dropNoise) out = out.split(WHITESPACE).filter(token => MEANINGFUL.test(token)).join(' ')
  if (o.stripPunctuation) out = out.replaceAll(PUNCTUATION, ' ')
  if (o.confusables) out = foldConfusables(out)

  return out.replaceAll(WHITESPACE, ' ').trim()
}

/** Normalised words, in order. */
export function tokenise (text: string, options: NormaliseOptions = {}): string[] {
  const normalised = normaliseText(text, options)

  return normalised === '' ? [] : normalised.split(' ')
}

File: python-packages/scanmate-ink/scanmate_ink/text_normalisation/normalise_text_algorithm.py
"""Text made comparable: the differences OCR introduces that say nothing about
the document are removed, and nothing else.

Each step is a judgement about which noise to forgive, so each can be turned
off. In order:

1. Unicode NFKC - ligatures, full-width and compatibility forms.
2. Typography - curly quotes, dashes, ellipses and odd spaces to plain ASCII.
3. De-hyphenation - a word broken across a line is rejoined, or every wrapped
   word counts as an edit.
4. Diacritics - accents folded away, ligatures split.
5. Case.
6. Noise - tokens with no letter, digit or currency sign: table rules read as
   pipes, underlines as runs of underscore, specks as a dot or a tilde.
7. Punctuation (off by default) - it carries real signal: a thousands
   separator and a decimal point are not noise.
8. OCR confusables (off by default) - folding them forgives exactly the
   substitutions a forger would make.
9. Whitespace collapsed and trimmed.

THREE PLACES WHERE PYTHON'S OWN TOOLS ARE THE WRONG TOOLS
----------------------------------------------------------

**Whitespace.** JavaScript's ``\\s`` and Python's are different sets. Python's
includes U+001C to U+001F and U+0085; JavaScript's does not. JavaScript's
includes U+FEFF; Python's does not. So this module carries the JavaScript class
explicitly rather than writing ``\\s``, and uses it for the final strip as
well - ``str.strip()`` with no argument would use Python's set again.

**Unicode properties.** ``\\p{L}``, ``\\p{N}``, ``\\p{Sc}`` and ``\\p{P}`` have
no equivalent in the standard ``re`` module. Rather than take a dependency on
``regex``, the three places that need them ask
:func:`unicodedata.category` directly, which is the same question the property
escape asks and needs no third-party code.

**Case.** ``str.lower()``, not ``str.casefold()``. Casefold is the aggressive
one - it turns the German sharp s into ``ss`` - and ``toLowerCase`` does not.
"""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, replace

from .confusables_mapper import fold_confusables
from .diacritics_mapper import fold_diacritics

#: JavaScript's ``\\s``: its WhiteSpace plus its LineTerminator. Written out
#: because Python's ``\\s`` is a different set in both directions.
_JS_SPACE = "\t\n\v\f\r \u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff"
_WHITESPACE = re.compile(f"[{_JS_SPACE}]+")

#: A letter, then a hyphen of some kind, then a line break. Whether the letter
#: really is a letter and what follows really is lower case is decided by
#: category in :func:`_rejoin`, since ``re`` cannot ask.
_LINE_END_HYPHEN = re.compile("(.)[-\u00ad\u2010\u2011][\t ]*\r?\n[\t ]*(?=(.))")

#: Curly quotes, dashes, ellipses and odd spaces, to plain ASCII. The order is
#: the TypeScript's and is kept for that reason alone - checked, and no rule
#: here produces a character another one consumes, so reversing them changes
#: nothing. That is worth writing down: the next reader should not have to
#: re-derive it, and should not assume there is a subtlety to preserve.
_TYPOGRAPHY: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile("[\u2018\u2019\u201a\u201b\u2032]"), "'"),
    (re.compile("[\u201c\u201d\u201e\u201f\u2033]"), '"'),
    (re.compile("[\u2010\u2011\u2012\u2013\u2014\u2015\u2212]"), "-"),
    (re.compile("\u2026"), "..."),
    (re.compile("[\u00a0\u2000-\u200a\u202f\u205f\u3000]"), " "),
    (re.compile("[\u200b\u200c\u2060\ufeff]"), ""),
    (re.compile("\u200d"), ""),
)

#: Unicode general categories that count as a letter, a number, or a currency sign.
_MEANINGFUL_CATEGORIES = frozenset({"Lu", "Ll", "Lt", "Lm", "Lo", "Nd", "Nl", "No", "Sc"})
#: Unicode general categories that count as punctuation.
_PUNCTUATION_CATEGORIES = frozenset({"Pc", "Pd", "Ps", "Pe", "Pi", "Pf", "Po"})


@dataclass(frozen=True, slots=True)
class NormaliseOptions:
    """Which of the nine steps run. The defaults are :data:`DEFAULT_NORMALISE`."""

    nfkc: bool = True
    typography: bool = True
    dehyphenate: bool = True
    diacritics: bool = True
    case_fold: bool = True
    drop_noise: bool = True
    #: Off by default: punctuation carries real signal in figures.
    strip_punctuation: bool = False
    #: Off by default: folding OCR's confusions hides what a check is for.
    confusables: bool = False


#: Every step at its default setting.
DEFAULT_NORMALISE = NormaliseOptions()


def _is_meaningful(token: str) -> bool:
    """Whether a token carries a letter, a digit or a currency sign.

    :param token: One whitespace-separated token.
    :returns: Whether anything in it is worth comparing.
    """
    return any(unicodedata.category(character) in _MEANINGFUL_CATEGORIES for character in token)


def _rejoin(match: re.Match[str]) -> str:
    """Drop a line-ending hyphen, but only between a letter and a lower-case one.

    The regex matched the shape; the categories decide whether it was really a
    wrapped word. A hyphen after a digit, or before a capital, is part of the
    text rather than an artefact of the line break.

    :param match: The candidate.
    :returns: The letter alone when it was a wrap, the whole match otherwise.
    """
    letter = match.group(1)
    following = match.group(2)
    if unicodedata.category(letter).startswith("L") and unicodedata.category(following) == "Ll":
        return letter

    return match.group(0)


def _strip_punctuation(text: str) -> str:
    """Replace every run of punctuation with a single space.

    A RUN, so an ellipsis rendered as three dots leaves one space rather than
    three. Transcribed that way because the TypeScript's pattern says so, not
    because it is observable: the final whitespace collapse squeezes the
    per-character version back to the same answer. Tried on 200,000 random
    strings - 131,473 differ before that collapse and none after it.

    :param text: The text.
    :returns: The text with punctuation runs replaced.
    """
    out: list[str] = []
    in_run = False
    for character in text:
        if unicodedata.category(character) in _PUNCTUATION_CATEGORIES:
            if not in_run:
                out.append(" ")
            in_run = True
        else:
            out.append(character)
            in_run = False

    return "".join(out)


def normalise_text(text: str, options: NormaliseOptions | None = None) -> str:
    """Run the enabled steps, in the order they are documented.

    :param text: The text to normalise.
    :param options: Which steps run.
    :returns: The normalised text.
    """
    chosen = options if options is not None else DEFAULT_NORMALISE
    out = text

    if chosen.nfkc:
        out = unicodedata.normalize("NFKC", out)
    if chosen.typography:
        for pattern, replacement in _TYPOGRAPHY:
            out = pattern.sub(replacement, out)
    if chosen.dehyphenate:
        out = _LINE_END_HYPHEN.sub(_rejoin, out)
    if chosen.diacritics:
        out = fold_diacritics(out)
    if chosen.case_fold:
        out = out.lower()
    if chosen.drop_noise:
        out = " ".join(token for token in _WHITESPACE.split(out) if _is_meaningful(token))
    if chosen.strip_punctuation:
        out = _strip_punctuation(out)
    if chosen.confusables:
        out = fold_confusables(out)

    # `strip()` with no argument would use PYTHON's whitespace set, and that is
    # a real difference rather than a defensive one: U+0085 is whitespace to
    # Python and not to JavaScript, so a bare `strip()` trims a leading or
    # trailing one that `trim()` leaves in place. The goldens carry that case.
    return _WHITESPACE.sub(" ", out).strip(" ")


def tokenise(text: str, options: NormaliseOptions | None = None) -> list[str]:
    """Normalised words, in order.

    :param text: The text to normalise.
    :param options: Which steps run.
    :returns: The words, or an empty list when nothing survived.
    """
    normalised = normalise_text(text, options)

    return [] if normalised == "" else normalised.split(" ")


def with_options(**overrides: bool) -> NormaliseOptions:
    """The defaults with some steps changed, mirroring a partial options object.

    The TypeScript spreads a partial object over the defaults; this is the same
    thing spelled for a frozen dataclass, so a caller need not restate the
    seven settings they are happy with.

    :param overrides: The steps to change.
    :returns: The resulting options.
    """
    return replace(DEFAULT_NORMALISE, **overrides)
